### PR TITLE
Allow public access for imports

### LINKED ISSUE #108776: Fix the imports in `torch.distributed._shard.sharding_spec.__init__.py`
### 🐛 Describe the bug

Similar to #80433, public access to `torch.distributed._shard.sharding_spec.ChunkShardingSpec` is not supported.

One possible fix (similar to #80955) is
```
from .chunk_sharding_spec import (
    ChunkShardingSpec as ChunkShardingSpec,
)
```

### Versions

[pip3] numpy==1.25.2
[pip3] torch==1.13.1+cu117
[pip3] torchtext==0.14.1
[pip3] torchvision==0.14.1+cu117
[conda] Could not collect

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu

### PR DESCRIPTION
**Pull request by b-chu**
Fixes #108776

### CODE REVIEW

File: torch/distributed/fsdp/__init__.py [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-    MixedPrecision,
-    OptimStateDictConfig,
-    OptimStateKeyType,
-    ShardedOptimStateDictConfig,
-    ShardedStateDictConfig,
-    ShardingStrategy,
-    StateDictConfig,
-    StateDictSettings,
-    StateDictType,
+    BackwardPrefetch as BackwardPrefetch,
```
Comment (wanchaol) [inline torch/distributed/fsdp/__init__.py:3]: hmmm wait, didn't expect a ton of changes like this as those seems not the right way to do. Could you describe your problem or do you have a specific repro?

Comment (wanchaol) [reply torch/distributed/fsdp/__init__.py:3]: Actually can we do `__all__` I think that would also enable public access

Comment (b-chu) [reply torch/distributed/fsdp/__init__.py]: Done.

**Comment by b-chu** [comment]
Are these test failures related to my change? How do I rerun tests to make sure it's not transient?

File: torch/distributed/fsdp/__init__.py [hunk 2]
```diff
@@ -17,3 +17,23 @@
     StateDictSettings,
     StateDictType,
 )
+
+__all__ = [
```
Comment (wanchaol) [inline torch/distributed/fsdp/__init__.py:21]: cc @awgu please review the FSDP part and see if there's anything we shouldn't expose

File: torch/distributed/_shard/sharding_spec/__init__.py [hunk 3]
```diff
@@ -6,7 +6,7 @@
     _infer_sharding_spec_from_shards_metadata,
 )
 from .chunk_sharding_spec import (
-    ChunkShardingSpec,
+    ChunkShardingSpec as ChunkShardingSpec,
```
Comment (fduwjj) [inline torch/distributed/_shard/sharding_spec/__init__.py:9]: For my learning purpose, why is this needed?

Comment (b-chu) [reply torch/distributed/_shard/sharding_spec/__init__.py:9]: This is for being able to import ChunkShardingSpec, otherwise public access isn't defined and linters complain, see #80433 for a similar issue.

Comment (fduwjj) [reply torch/distributed/_shard/sharding_spec/__init__.py:9]: Got it...It's just the way the code looks like is bit weird to me.

**Comment by b-chu** [comment]
@wanchaol @awgu Are there additional changes needed?

File: torch/distributed/fsdp/__init__.py [hunk 4]
```diff
@@ -17,3 +17,23 @@
...
+    "FlatParameter",
```
Comment (wanchaol) [inline torch/distributed/fsdp/__init__.py:22]: similar comment from @fegin  in https://github.com/pytorch/pytorch/pull/109922, we should not expose `FlatPrameter` as a public API, besides that the rest looks good to me

**Review by wanchaol — APPROVED** [review]
lgtm, thanks!
